Feat: support realm agnostic migrations (#176)
witnet/witnet-ethereum-bridge: README.md

# witnet-ethereum-bridge [![](https://travis-ci.com/witnet/witnet-ethereum-bridge.svg?branch=master)](https://travis-ci.com/witnet/witnet-ethereum-bridge)

> `witnet/witnet-ethereum-bridge` is an open source implementation of an API that enable Solidity smart contract developers to harness the full power of the [**Witnet Decentralized Oracle Network**](https://docs.witnet.io/overview/concepts/).

Reviewer: ```suggestion
`witnet/witnet-ethereum-bridge` is an open source implementation of an API that enables Solidity smart contract developers to harness the full power of the [**Witnet Decentralized Oracle Network**](https://docs.witnet.io/overview/concepts/).
```

"the full power"
![image](https://user-images.githubusercontent.com/12464544/128492915-08289218-7df7-4e8e-aa98-3aa6f79e2cc8.png)

This repository provides several deployable contracts:

- `WitnetProxy`, that routes Witnet data requests to a currently active `WitnetRequestBoard` implementation.
- `WitnetRequest`, used as a means to encapsulate CBOR-encoded Witnet RADON scripts.
- `WitnetRequestBoard` (WRB), which implements all required functionality to relay Witnet data requests (i.e. encapsulated [Witnet RADON scripts](https://docs.witnet.io/protocol/data-requests/overview/)) from Ethereum to the Witnet mainnet, as well as to relay Witnet-solved results back to Ethereum.

> - `UsingWitnet`, an inheritable abstract contract that injects methods for interacting with the WRB in the most convenient way.

Reviewer: ```suggestion
- `UsingWitnet`, an inheritable abstract contract that injects methods for conveniently interacting with the WRB.
```

Maybe the `UsingWitnet` should not be on this list as by itself is not deployeable (abstract contract).

Author: ```
- `WitnetRequestBoard` (WRB), which implements all required functionality to relay Witnet data requests (i.e. encapsulated [Witnet RADON scripts](https://docs.witnet.io/protocol/data-requests/overview/)) from Ethereum to the Witnet mainnet, as well as to relay Witnet-solved results back to Ethereum.

The repository also provides:

- `UsingWitnet`, an inheritable abstract contract that injects methods for conveniently interacting with the WRB.
```

## WitnetProxy

`WitnetProxy` is an upgradable delegate-proxy contract that routes Witnet data requests coming from a `UsingWitnet`-inheriting contract to a currently active `WitnetRequestBoard` implementation. 

> This table contains all the `WinetProxy` instances that act as entry-points for the latest versions of the `WitnetRequestBoard` approved and deployed by the [**Witnet Foundation**](https://witnet.io), and that actually integrate with the **Witnet mainnet** from the following EVM-compatible blockchains:

Reviewer: ```suggestion
This table contains all the `WinetProxy` instances that act as entry-points for the latest versions of the `WitnetRequestBoard` approved and deployed by the [**Witnet Foundation**](https://witnet.io), and that actually integrates with the **Witnet mainnet** from the following EVM-compatible blockchains:
```

  | Blockchain   | Network   | `WitnetProxy` address
  | ------------ | --------- | ---------------------
  | **Ethereum** | Rinkeby   | `` 
  |              | Göerli    | `` 
  |              | Mainnet   | `` 
  | ------------ | --------- | ---------------------
  | **Conflux**  | Testnet   | `` 
  |              | Mainnet   | `` 
  | ------------ | --------- | ---------------------

## WitnetRequest

>  A `WitnetRequest` is constructed around a `bytes` value containing a well-formed Witnet RADON script data request serialized
>  using Witnet own protocol buffers. The `WitnetRequest` base contract provides one single method:

Reviewer 2: ```suggestion
A `WitnetRequest` is constructed around a `bytes` value containing a well-formed Witnet RADON script data request serialized
using Protocol Buffers. The `WitnetRequest` base contract provides one single method:
```

- **`bytecode()`**:
  - _Description_:
    - Returns Witnet RADON script as a CBOR-encoded `bytes`. 

## WitnetRequestBoard

The `WitnetRequestBoard` implements the following functionality:

- **`estimateGasCost(uint256 _gasPrice)`**:
  - _Description_: 

>     - Estimates the minimal amount of reward needed as to post a Witnet data request into the WRB, for a given gas price.

Reviewer: ```suggestion
    - Estimates the minimal amount of reward needed to post a Witnet data request into the WRB, for a given gas price.
```

  - _Inputs_:
    - `_gasPrice`: the request contract address which includes the Witnet RADON script bytecode.
  - _Returns_:
    - The minimal reward amount. 

> - **`destoryResult(uint256 _id)`**:

Reviewer 2: ```suggestion
- **`destroyResult(uint256 _id)`**:
```

  - _Description_:
    - Retrieves the Witnet-solved result (if already available) of a previously posted Witnet request, and removes it from the WRB storage.
  - _Inputs_:
    - `_id`: the unique identifier of a previously posted Witnet data request.
  - _Returns_:
    - The Witnet-solved result of the given data request, as CBOR-encoded `bytes`.
  
- **`postDataRequest(address _witnetRequest)`**:
  - _Description_: 

>     - Posts a Witnet data request into the WRB in expectation that it will be eventually relayed and resolved 

Reviewer: ```suggestion
    - Posts a Witnet data request into the WRB in the expectation that it will be eventually relayed and resolved 
```

  by Witnet, with `msg.value` as reward.
  - _Inputs_:
    - `_witnetRequest`: the actual `WitnetRequest` contract address which provided the Witnet RADON script bytecode.
  - _Returns_:
    - *_id*: the unique identifier of the data request.

- **`readDataRequest(uint256 _id)`**:
  - _Description_:

>     - Retrieves the RADON script bytcode of a previously posted Witnet data request.

Reviewer: ```suggestion
    - Retrieves the RADON script bytecode of a previously posted Witnet data request.
```

  - _Description_: 
    - Checks if a Witnet data request has already been soled by Witnet network.
  - _Inputs_:
    - `_id`: the unique identifier of a previously posted Witnet data request.
  - _Returns_:
    - A boolean telling whether the Witnet data request has been already solved, or not.

- **`witnetReadResult(uint256 _id)`**:
  - _Description_:
    - Retrieves the Witnet-solved result of a previously posted Witnet data request.
  - _Inputs_:
    - `_id`: the unique identifier of a previously posted Witnet data request.
  - _Returns_:
    - The Witnet-solved result of the given data request, as CBOR-encoded `bytes`.

- **`witnetDestroyResult(uint256 _id)`**:
  - _Description_:
    - Retrieves the Witnet-solved result (if already available) of a previously posted Witnet request, and removes it from the WRB storage. Works only if this `UsingWitnet` contract is the one that actually posted the given data request.
  - _Inputs_:
    - `_id`: the unique identifier of a previously posted Witnet data request.
  - _Returns_:
    - The Witnet-solved result of the given data request, as CBOR-encoded `bytes`.

## Usage: harness the power of the Witnet Decentralized Oracle Network

> In order to integrate your own smart contracts with the **Witnet** fully-decentalized blockchain, you just need to inherit from the `UsingWitnet` abstract contract:

Reviewer: ```suggestion
In order to integrate your own smart contracts with the **Witnet** fully-decentralized blockchain, you just need to inherit from the `UsingWitnet` abstract contract:
```

```solidity
pragma solidity >=0.7.0 <0.9.0;

import "./UsingWitnet.sol";

contract MyWitnetRequest is WitnetRequest {
  constructor() WitnetRequest()
}

contract MyContract is UsingWitnet {
  WitnetRequest myRequest;

  constructor() UsingWitnet(/* WitnetProxy address provided by the Witnet Foundation */) {
    myRequest = new WitnetRequest(hex"/* here goes the Witnet RADON script encoded as serialized bytes. */")
  }

  function myOwnDrPost() public returns(uint256 _drTrackId) {
    _drTrackId = witnetPostRequest{value: msg.value}(myRequest);
  }
}
```

> Please, have a look to the [`witnet/witnet-price-feed-examples`](https://github.com/witnet/witnet-price-feed-examples) repository to learn how to compose your own `WitnetRequest` contracts.

Reviewer: ```suggestion
Please, have a look at the [`witnet/witnet-price-feed-examples`](https://github.com/witnet/witnet-price-feed-examples) repository to learn how to compose your own `WitnetRequest` contracts.
```
